File: problems/0981-time-based-key-value-store/typescript/solution.ts
// Time Based Key-Value Store — https://leetcode.com/problems/time-based-key-value-store/
class TimeMap {
  private store = new Map<string, { timestamp: number; value: string }[]>();

  set(key: string, value: string, timestamp: number): void {
    const entries = this.store.get(key);
    if (entries) entries.push({ timestamp, value });
    else this.store.set(key, [{ timestamp, value }]);
  }

  get(key: string, timestamp: number): string {
    const entries = this.store.get(key) ?? [];
    let lo = 0;
    let hi = entries.length - 1;
    let res = "";
    while (lo <= hi) {
      const mid = Math.floor((lo + hi) / 2);
      if (entries[mid].timestamp <= timestamp) {
        res = entries[mid].value;
        lo = mid + 1;
      } else {
        hi = mid - 1;
      }
    }
    return res;
  }
}

// Test driver: replay LeetCode-style operation/argument lists (null for void ops).
export function timeMapOps(ops: string[], args: (string | number)[][]): (string | null)[] {
  const tm = new TimeMap();
  const out: (string | null)[] = [];
  for (let i = 0; i < ops.length; i++) {
    const arg = args[i];
    switch (ops[i]) {
      case "TimeMap":
        out.push(null);
        break;
      case "set":
        tm.set(String(arg[0]), String(arg[1]), Number(arg[2]));
        out.push(null);
        break;
      case "get":
        out.push(tm.get(String(arg[0]), Number(arg[1])));
        break;
    }
  }
  return out;
}

File: problems/0981-time-based-key-value-store/go/solution.go
package solution

// Time Based Key-Value Store — https://leetcode.com/problems/time-based-key-value-store/
type tmEntry struct {
	timestamp int
	value     string
}

type TimeMap struct {
	store map[string][]tmEntry
}

func Constructor() TimeMap {
	return TimeMap{store: make(map[string][]tmEntry)}
}

func (m *TimeMap) Set(key, value string, timestamp int) {
	m.store[key] = append(m.store[key], tmEntry{timestamp, value})
}

func (m *TimeMap) Get(key string, timestamp int) string {
	entries := m.store[key]
	lo, hi := 0, len(entries)-1
	res := ""
	for lo <= hi {
		mid := lo + (hi-lo)/2
		if entries[mid].timestamp <= timestamp {
			res = entries[mid].value
			lo = mid + 1
		} else {
			hi = mid - 1
		}
	}
	return res
}

// timeMapOps is the test driver: replay LeetCode-style operation/argument lists.
// args carry mixed string/number values, so they decode as [][]any.
func timeMapOps(ops []string, args [][]any) []*string {
	tm := Constructor()
	out := []*string{}
	for i, op := range ops {
		switch op {
		case "TimeMap":
			out = append(out, nil)
		case "set":
			key := args[i][0].(string)
			value := args[i][1].(string)
			ts := int(args[i][2].(float64))
			tm.Set(key, value, ts)
			out = append(out, nil)
		case "get":
			key := args[i][0].(string)
			ts := int(args[i][1].(float64))
			v := tm.Get(key, ts)
			out = append(out, &v)
		}
	}
	return out
}
